examples/homogeneous/README.md

```diff
@@ -0,0 +1,204 @@
+## Contract invariants - model checking and proof assistant methods
+
+The majority of the formal verification work that has been employed in practice so far in the ethereum ecosystem has been dealing with _functional correctness_, that is, given a specification of how a contract method should update the state, or an explicit assertion in the body of a method, demonstrate that the implementation correctly matches this specification or satisfies the assertion made. This can be done on an EVM level using tools like [KEVM](https://www.ideals.illinois.edu/handle/2142/97207) or on the solidity level using the [builtin smt-checker](https://github.com/leonardoalt/text/blob/master/solidity_isola_2018/main.pdf) or [solc-verify](https://arxiv.org/abs/1907.04262).
+
+But often, the most crucial elements of how the smart contract behaves arises out of the interplay between its different functions, and its essence is best captured through a set of _invariant properties_, i.e. a predicate over its state variables which should hold before and after every function call. Some tools can already prove simple invariants, but struggle with invariants containing non-linear arithmetic expressions. 
+
+## Linear invariants
+
+Here is an example of a invariant provable by solc-verify
+(solc-verfiy)
+```sol
+pragma solidity >=0.5.0;
+
+/**
+ * @notice invariant __verifier_sum_uint(balances) <= address(this).balance
+ */
+contract SimpleBank {
+    mapping(address=>uint) balances;
+
+    function deposit() public payable {
+        balances[msg.sender] += msg.value;
+    }
+
+    function withdraw(uint256 amount) public {
+        require(balances[msg.sender] > amount);
+        balances[msg.sender] -= amount;
+        bool ok;
+        (ok, ) = msg.sender.call.value(amount)(""); // No reentrancy attack
+        if (!ok) revert();
+    }
+}
+```
+
+and here is an example by the smt-checker:
+```sol
+pragma experimental SMTChecker;
+
+contract C {
+	uint x;
+
+	function f() public {
+		if (x == 0)
+			x = 1;
+	}
+
+	function g() public {
+		if (x == 1)
+			x = 2;
+	}
+
+	function h() public {
+		if (x == 2)
+			x = 0;
+	}
+
+	// This function shows that (x < 9) is not inductive and
+	// a stronger invariant is needed to be found.
+	// (x < 3) is the one found in the end.
+	function j() public {
+		if (x == 7)
+			x = 100;
+	}
+
+	function i() public view {
+		assert(x < 9);
+	}
+}
+```
+
+The smt-checker can in some cases also automatically infer invariants. 
+
+The examples above are easy for SMT-solvers to reason about, since they deal with expressions of [linear arithmetic](https://cse-wiki.unl.edu/wiki/images/0/04/DecisionProcedure-Chapter5a.pdf). Linear arithmetic expressions can contain any valid combination of the symbols `<, <=, and, or, ==, +, -`, but they can crucially not contain multiplication by a an unknown variable.
```

Comment:
```suggestion
The examples above are easy for SMT solvers to reason about, since they deal with expressions of [linear arithmetic](https://cse-wiki.unl.edu/wiki/images/0/04/DecisionProcedure-Chapter5a.pdf). Linear arithmetic expressions can contain any valid combination of the symbols `<, <=, and, or, ==, +, -`, but they can crucially not contain multiplication between variables.
```